### PR TITLE
fix clang animation bug

### PR DESCRIPTION
**Pull request by marklemay**

### CODE REVIEW

**Comment by hifi** [comment]
If this indeed fixes clang builds it needs to be done for all `TFixedIHeapClass` implementations that inhibit the same style of constructor:

https://github.com/TheAssemblyArmada/Vanilla-Conquer/blob/vanilla/redalert/externs.h#L205

I'm afraid if it's not done properly this will creep up again when any other heap limit is hit.

**Comment by marklemay** [comment]
I think there is a way to turn on a warning for where this should be used.  we can be more systematic about all the places it should go

**Comment by giulianobelinassi** [comment]
Running
```
$ grep -rnI "operator new" $(find -name "*.h")
```

gives me

```
./tiberiandawn/terrain.h:57:    static void* operator new(size_t size);  
./tiberiandawn/terrain.h:58:    static void* operator new(size_t, void* ptr)                                                                                                         [50/1904]
./tiberiandawn/overlay.h:52:    static void* operator new(size_t size);
./tiberiandawn/overlay.h:53:    static void* operator new(size_t, void* ptr)
./tiberiandawn/teamtype.h:144:    void* operator new(size_t);     
./tiberiandawn/teamtype.h:145:    static void* operator new(size_t, void* ptr)
./tiberiandawn/factory.h:49:    static void* operator new(size_t size);
./tiberiandawn/factory.h:50:    static void* operator new(size_t, void* ptr)
./tiberiandawn/team.h:189:    static void* operator new(size_t, void* ptr)
./tiberiandawn/team.h:193:    static void* operator new(size_t size); 
./tiberiandawn/bullet.h:69:    static void* operator new(size_t size);
./tiberiandawn/bullet.h:70:    static void* operator new(size_t, void* ptr)
./tiberiandawn/smudge.h:53:    static void* operator new(size_t size);
./tiberiandawn/smudge.h:54:    static void* operator new(size_t, void* ptr)
./tiberiandawn/house.h:414:    static void* operator new(size_t size);
./tiberiandawn/house.h:415:    static void* operator new(size_t, void* ptr)
./tiberiandawn/house.h:913:        static void* operator new(size_t, void* ptr)
./tiberiandawn/trigger.h:198:    static void* operator new(size_t size);
./tiberiandawn/trigger.h:199:    static void* operator new(size_t, void* ptr)
./tiberiandawn/infantry.h:115:    static void* operator new(size_t size);
./tiberiandawn/infantry.h:116:    static void* operator new(size_t, void* ptr)
./tiberiandawn/aircraft.h:51:    void* operator new(size_t);           
./tiberiandawn/aircraft.h:53:    static void* operator new(size_t, void* ptr)
./tiberiandawn/building.h:165:    static void* operator new(size_t size);
./tiberiandawn/building.h:166:    static void* operator new(size_t, void* ptr)
./tiberiandawn/unit.h:60:    static void* operator new(size_t size);   
./tiberiandawn/unit.h:61:    static void* operator new(size_t, void* ptr)
./tiberiandawn/anim.h:48:    static void* operator new(size_t size);   
./tiberiandawn/anim.h:49:    static void* operator new(size_t, void* ptr)
./tiberiandawn/template.h:52:    static void* operator new(size_t size);
./tiberiandawn/template.h:53:    static void* operator new(size_t, void* ptr)
./common/memflag.h:46:    MEM_NEW = 0x0001,    // Called by the operator new and was overloaded.
./common/memflag.h:62:void* operator new(size_t size, MemoryFlagType flag);
./common/memflag.h:63:void* operator new[](size_t size, MemoryFlagType flag);
./common/memflag.h:74:inline void* operator new(size_t size, MemoryFlagType flag)
./common/memflag.h:78:inline void* operator new[](size_t size, MemoryFlagType flag)
./redalert/vessel.h:81:    static void* operator new(size_t size);     
./redalert/vessel.h:82:    static void* operator new(size_t, void* ptr)
./redalert/terrain.h:57:    static void* operator new(size_t size);       
./redalert/terrain.h:58:    static void* operator new(size_t, void* ptr)
./redalert/trigtype.h:108:    static void* operator new(size_t);         
./redalert/trigtype.h:109:    static void* operator new(size_t, void* ptr)
./redalert/overlay.h:57:    static void* operator new(size_t size);       
./redalert/overlay.h:58:    static void* operator new(size_t, void* ptr)
./redalert/teamtype.h:128:    static void* operator new(size_t);      
./redalert/teamtype.h:129:    static void* operator new(size_t, void* ptr)
./redalert/factory.h:50:    static void* operator new(size_t size);  
./redalert/factory.h:51:    static void* operator new(size_t, void* ptr)
./redalert/warhead.h:50:    void* operator new(size_t);             
./redalert/warhead.h:51:    static void* operator new(size_t, void* ptr) 
./redalert/team.h:210:    static void* operator new(size_t size);  

./redalert/team.h:211:    static void* operator new(size_t, void* ptr)
./redalert/bullet.h:67:    static void* operator new(size_t size);
./redalert/bullet.h:68:    static void* operator new(size_t, void* ptr)
./redalert/smudge.h:58:    static void* operator new(size_t size);
./redalert/smudge.h:59:    static void* operator new(size_t, void* ptr)
./redalert/type.h:178:    static void* operator new(size_t);
./redalert/type.h:179:    static void* operator new(size_t, void* ptr)
./redalert/type.h:820:    static void* operator new(size_t);
./redalert/type.h:821:    static void* operator new(size_t, void* ptr)
./redalert/type.h:1035:    static void* operator new(size_t);
./redalert/type.h:1036:    static void* operator new(size_t, void* ptr)
./redalert/type.h:1145:    static void* operator new(size_t);
./redalert/type.h:1146:    static void* operator new(size_t, void* ptr)
./redalert/type.h:1304:    static void* operator new(size_t);
./redalert/type.h:1305:    static void* operator new(size_t, void* ptr)
./redalert/type.h:1410:    static void* operator new(size_t);
./redalert/type.h:1411:    static void* operator new(size_t, void* ptr)
./redalert/type.h:1597:    static void* operator new(size_t);
./redalert/type.h:1598:    static void* operator new(size_t, void* ptr)
./redalert/type.h:1669:    static void* operator new(size_t);
./redalert/type.h:1670:    static void* operator new(size_t, void* ptr)
./redalert/type.h:1732:    static void* operator new(size_t);
./redalert/type.h:1733:    static void* operator new(size_t, void* ptr)
./redalert/type.h:1961:    static void* operator new(size_t);
./redalert/type.h:1962:    static void* operator new(size_t, void* ptr)
./redalert/type.h:2081:    static void* operator new(size_t);
./redalert/type.h:2082:    static void* operator new(size_t, void* ptr)
./redalert/type.h:2157:    static void* operator new(size_t);
./redalert/type.h:2158:    static void* operator new(size_t, void* ptr)
./redalert/house.h:654:    static void* operator new(size_t size);
./redalert/house.h:655:    static void* operator new(size_t, void* ptr)
./redalert/house.h:943:        static void* operator new(size_t, void* ptr)
./redalert/trigger.h:95:    static void* operator new(size_t size);
./redalert/trigger.h:96:    static void* operator new(size_t, void* ptr)
./redalert/weapon.h:52:    void* operator new(size_t);
./redalert/weapon.h:53:    static void* operator new(size_t, void* ptr)
./redalert/infantry.h:112:    static void* operator new(size_t size);
./redalert/infantry.h:113:    static void* operator new(size_t, void* ptr)
./redalert/aircraft.h:59:    static void* operator new(size_t);
./redalert/aircraft.h:60:    static void* operator new(size_t, void* ptr)
./redalert/building.h:213:    static void* operator new(size_t size);
./redalert/building.h:214:    static void* operator new(size_t, void* ptr)
./redalert/unit.h:128:    static void* operator new(size_t size);
./redalert/unit.h:129:    static void* operator new(size_t, void* ptr)
./redalert/anim.h:65:    static void* operator new(size_t size);
./redalert/anim.h:66:    static void* operator new(size_t, void* ptr)
./redalert/compat.h:177:    static void* operator new(size_t);
./redalert/template.h:57:    static void* operator new(size_t size);
./redalert/template.h:58:    static void* operator new(size_t, void* ptr)
```

Perhaps you can craft a `sed` regex that fixes everything in a single command.

**Review by OmniBlade — COMMENTED** [review]
The operators of the form new(size_t, void* ptr) shouldn't need to be marked as noexcept as they are placement new and assume the caller has already allocated the memory so they just need to invoke the constructor.

**Comment by MLipfert1** [comment]
On my forked branch, I have created a commit that adds "noexcept" to all signatures of type new(size_t). Please see [here](https://github.com/MLipfert1/Vanilla-Conquer/commit/2f994515d9b26de6eedbc79280073f287f60f5ca).

This seems to indeed fix the animation crash, as I was able to finish the first mission of the campaign for the first time without exception.

Do we want to go on with the small commit from marklemay or do we really want to touch every occurrence now? If we want to go on with the big solution, my next question would be, should I create a new pull request or do you, marklemay, want to do it, because you initially found the solution?

File: redalert/anim.h [hunk 1]
```diff
@@ -62,8 +62,8 @@ class AnimClass : public ObjectClass, public StageClass
         return Class->Type;
     };
 
-    static void* operator new(size_t size);
-    static void* operator new(size_t, void* ptr)
+    static void* operator new(size_t size) noexcept;
+    static void* operator new(size_t, void* ptr) noexcept
```
Comment (OmniBlade) [inline redalert/anim.h]: This isn't needed as this is just placement new that has to be passed a pointer to the data to run the constructor on anyway.

**Review by OmniBlade — CHANGES_REQUESTED** [review]
If you can revert the placement new change I will merge this and we can look at a larger PR to change all the other heaps as well. Once that is done we can probably remove the -fcheck-new option from the compiler flags as well.

**Comment by OmniBlade** [comment]
@marklemay If you don't mind updating the PR to remove the unneeded change to placement new I will merge this as is as it fixed and immediate problem. If not then I'm happy for @MLipfert1 to PR a change to update all the overridden `operator new` instances to replace this PR.

**Comment by marklemay** [comment]
@MLipfert1 I'm happy to use your commit.  I had planned on running clang-tidy and only marking the operators with `-Wnew-returns-null` but I don't see why adding `noexcept` everywhere would be a problem.

Also I wanted to add a small test that over allocates and confirms the expected behavior, since it is a little unusual and may come up again if someone tries to build with a new compiler. 

That said, I might be busier than I thought.  So feel free to handle this however you want.

File: redalert/anim.h (hunk 1 above)
Comment (marklemay) [reply redalert/anim.h]: sounds good

**Comment by MLipfert1** [comment]
I created a PR for the full commit that adds noexcept to all operator new signatures mentioned above.
See #817 .

**Comment by OmniBlade** [comment]
@marklemay I'll close this PR in favour of @MLipfert1 one then, though it would be good to get a test case included if you have the time to submit one in the future.

**Pull request closed without merging**